contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+Scheduler stops running the remaining filter functions for a node once one of
+these filters fails for the node.
+
+### Post-filter
+
+The Pod and the set of nodes that can run the Pod are passed to these plugins.
+They are called whether Pod is schedulable or not (whether the set of nodes is
+empty or non-empty).  
+If any of these plugins return an error or if the Pod is determined
+unschedulable, the scheduling cycle is aborted.  
+These plugins are called serially.
+
+### Scoring
+
+These plugins are similar to priority function in scheduler v1. They are
+utilized to rank nodes that have passed the filtering stage. Similar to Filter
+plugins, these are called per node serially in the same order registered, but
+scheduler may run them for multiple nodes in parallel.  
+Each one of these functions return a score for the given node. The score is
+multiplied by the weight of the function and aggregated with the result of other
+scoring functions to yield a total score for the node.  
+These functions can never block scheduling. In case of an error they should
+return zero for the Node being ranked.
+
+### Post-scoring/pre-reservation
+
+After all scoring plugins are invoked and the score of nodes are determined, the
+framework picks the best node with the highest score and then it calls
+post-scoring plugins. The Pod and the chosen Node are passed to these plugins.
+These plugins have one more chance to check any conditions about the assignment
+of the Pod to this Node and reject the node if needed.
+
+![image](images/scheduling-framework-threads.png)
+
+### Reserve
+
+This is not a plugin point. At this point scheduler updates its cache by
+"reserving" a Node (partially or fully) for the Pod. In scheduler v1 this stage
+is called "assume". At this point, only the scheduler cache is updated to
+reflect that the Node is (partially) reserved for the Pod. The actual assignment
+of the Node to the Pod happens during the "Bind" phase. That is when the API
+server updates the Pod object with the Node information.
+
+### Admit
+
+Admit plugins run in a separate go routine (in parallel). Each plugin can return
+one of the three possible values: 1) "admit", 2) "reject", or 3) "wait". If all
+plugins registered at this extension point return "admit", the pod is sent to
+the next step for binding. If any of the plugins returns "reject", the pod is
+rejected and sent back to the scheduling queue. If any of the plugins returns
+"wait", the Pod is kept in reserved state until it is explicitly approved for
+binding. A plugin that returns "wait" must return a "timeout" as well. If the
+timeout expires, the pod is rejected and goes back to the scheduling queue.
+
+#### Approving a Pod binding
+
+While any plugin can receive the list of reserved Pod from the cache and approve
+them,  we expect only the "Admit" plugins to approve binding of reserved Pods
+that are in "waiting" state. Once a Pod is approved, it is sent to the Bind
+stage.
+
+### Reject
```

Comment:
Why can't we combine this into admit extension point, when there is  a failure, we can undo all the activities done till now.

Reply:
The reason is that "admit" may have multiple plugins. They are called in order. Let's say the last one fails, while all the other plugins have already performed their tasks. There must be another extension point that allows those plugins to undo their operations when needed.